Permission created before making device a control node is not getting active

### Description

1. recover account that is an owner of any community
2. enable test nets
3. enable managing communities on test nets
4. open community - permssions - create permission
5. note that permission is created with status Pending
6. make the device a control node
7. restart the app

As result, the permission is never active even the device is now a control node

<img width="1589" alt="Screenshot 2024-03-19 at 11 01 43" src="https://github.com/status-im/status-desktop/assets/82375995/6f08a0ae-4190-42ca-8e23-3ee5aa378e63">

On master, attempt to reproduce the problem : 

I followed steps 1 - 6, which ended in a crash. Trying to start the app crashes constantly when setting the device as control node.

```bash
INF 2024-03-20 23:57:04.134-07:00 starting application...                    topics="status-app" tid=659651 file=nim_status_client.nim:205
WRN 2024-03-20 23:57:04.134-07:00 Error decoding signal                      topics="signals-manager" tid=659651 file=signals_manager.nim:49 err="Unknown signal received: mediaserver.started"
DBG 2024-03-20 23:57:08.253-07:00 primary_action                             topics="app-controller" tid=659651 file=module.nim:207 currFlow=AppLogin currState=LoginKeycardEnterPassword
ERROR[03-20|23:57:08.562|github.com/status-im/status-go/api/geth_backend.go:430] failed to initialize wallet db           package=status-go/api.GethStatusBackend err="failed to set `journal_mode` pragma: file is not a database"
failed to set `journal_mode` pragma: file is not a database
ERR 2024-03-20 23:57:08.562-07:00 error:                                     topics="accounts-service" tid=659651 file=service.nim:656 procName=verifyDatabasePassword errDesription="failed to set `journal_mode` pragma: file is not a database"
DBG 2024-03-20 23:57:08.563-07:00 Account logged in                          topics="accounts-service" tid=659651 file=service.nim:676
WARN [03-20|23:57:09.152|github.com/status-im/status-go/api/geth_backend.go:716] fleet is not supported, overriding with default value package=status-go/api.GethStatusBackend fleet= defaultFleet=shards.test
WRN 2024-03-20 23:57:09.167-07:00 Error decoding signal                      topics="signals-manager" tid=659651 file=signals_manager.nim:49 err="Unknown signal received: mediaserver.started"
SIGSEGV: Illegal storage access. (Attempt to read from nil?)
make: *** [Makefile:815: run-linux] Segmentation fault (core dumped)
```

I will try to

--- a/protocol/communities/manager.go
+++ b/protocol/communities/manager.go
@@ -2107,17 +2107,9 @@ func (m *Manager) HandleCommunityEventsMessage(signer *ecdsa.PublicKey, message 
 			return nil, err
 		}
 	}
-	err = community.processEvents(eventsMessage, lastlyAppliedEvents)
-	if err != nil {
-		return nil, err
-	}
 
-	additionalCommunityResponse, err := m.handleAdditionalAdminChanges(community)
+	additionalCommunityResponse, err := m.handleCommunityEventsAndMetadata(community, eventsMessage, lastlyAppliedEvents)
 	if err != nil {
-		return nil, err
-	}
-
-	if err = m.handleCommunityTokensMetadata(community); err != nil {
 		return nil, err
 	}
 
@@ -5289,9 +5281,73 @@ func (m *Manager) promoteSelfToControlNode(community *Community, clock uint64) (
 		return false, err
 	}
 
+	err = m.handleCommunityEvents(community)
+	if err != nil {
+		return false, err
+	}
+
 	community.increaseClock()
 
 	return ownerChanged, nil
+}
+
+func (m *Manager) handleCommunityEventsAndMetadata(community *Community, eventsMessage *CommunityEventsMessage,
+	lastlyAppliedEvents map[string]uint64) (*CommunityResponse, error) {
+	err := community.processEvents(eventsMessage, lastlyAppliedEvents)
+	if err != nil {
+		return nil, err
+	}
+
+	additionalCommunityResponse, err := m.handleAdditionalAdminChanges(community)
+	if err != nil {
+		return nil, err
+	}
+
+	if err = m.handleCommunityTokensMetadata(community); err != nil {
+		return nil, err
+	}
+
+	return additionalCommunityResponse, err
+}
+
+func (m *Manager) handleCommunityEvents(community *Community) error {
+	if community.config.EventsData == nil {
+		return nil
+	}
+
+	lastlyAppliedEvents, err := m.persistence.GetAppliedCommunityEvents(community.ID())
+	if err != nil {
+		return err
+	}
+
+	_, err = m.handleCommunityEventsAndMetadata(community, community.toCommunityEventsMessage(), lastlyAppliedEvents)
+	if err != nil {
+		return err
+	}
+
+	appliedEvents := map[string]uint64{}
+	if community.config.EventsData != nil {
+		for _, event := range community.config.EventsData.Events {
+			appliedEvents[event.EventTypeID()] = event.CommunityEventClock
+		}
+	}
+
+	community.config.EventsData = nil // clear events, they are already applied
+	community.increaseClock()
+
+	err = m.persistence.SaveCommunity(community)
+	if err != nil {
+		return err
+	}
+
+	err = m.persistence.UpsertAppliedCommunityEvents(community.ID(), appliedEvents)
+	if err != nil {
+		return err
+	}
+
+	m.publish(&Subscription{Community: community})
+
+	return nil
 }
 
 func (m *Manager) shareRequestsToJoinWithNewPrivilegedMembers(community *Community, newPrivilegedMembers map[protobuf.CommunityMember_Roles][]*ecdsa.PublicKey) error {

--- a/protocol/communities_messenger_signers_test.go
+++ b/protocol/communities_messenger_signers_test.go
@@ -732,3 +732,96 @@ func (s *MessengerCommunitiesSignersSuite) TestSyncTokenGatedCommunity() {
 		})
 	}
 }
+
+func (s *MessengerCommunitiesSignersSuite) TestWithMintedOwnerTokenApplyCommunityEventsUponMakingDeviceControlNode() {
+	community := s.createCommunity(s.john)
+
+	// john mints owner token
+	var chainID uint64 = 1
+	tokenAddress := "token-address"
+	tokenName := "tokenName"
+	tokenSymbol := "TSM"
+	_, err := s.john.SaveCommunityToken(&token.CommunityToken{
+		TokenType:       protobuf.CommunityTokenType_ERC721,
+		CommunityID:     community.IDString(),
+		Address:         tokenAddress,
+		ChainID:         int(chainID),
+		Name:            tokenName,
+		Supply:          &bigint.BigInt{},
+		Symbol:          tokenSymbol,
+		PrivilegesLevel: token.OwnerLevel,
+	}, nil)
+	s.Require().NoError(err)
+
+	err = s.john.AddCommunityToken(community.IDString(), int(chainID), tokenAddress)
+	s.Require().NoError(err)
+
+	// Make sure there is no control node
+	s.Require().False(common.IsPubKeyEqual(community.ControlNode(), &s.john.identity.PublicKey))
+
+	// Trick. We need to remove the community private key otherwise the events
+	// will be signed and Events will be approved instead of being in Pending State.
+	_, err = s.john.RemovePrivateKey(community.ID())
+	s.Require().NoError(err)
+
+	request := requests.CreateCommunityTokenPermission{
+		CommunityID: community.ID(),
+		Type:        protobuf.CommunityTokenPermission_BECOME_ADMIN,
+		TokenCriteria: []*protobuf.TokenCriteria{
+			&protobuf.TokenCriteria{
+				Type:              protobuf.CommunityTokenType_ERC20,
+				ContractAddresses: map[uint64]string{testChainID1: "0x123"},
+				Symbol:            "TEST",
+				AmountInWei:       "100000000000000000000",
+				Decimals:          uint64(18),
+			},
+		},
+	}
+
+	response, err := s.john.CreateCommunityTokenPermission(&request)
+	s.Require().NoError(err)
+	s.Require().Len(response.CommunityChanges, 1)
+	s.Require().Len(response.CommunityChanges[0].TokenPermissionsAdded, 1)
+
+	addedPermission := func() *communities.CommunityTokenPermission {
+		for _, permission := range response.CommunityChanges[0].TokenPermissionsAdded {
+			return permission
+		}
+		return nil
+	}()
+	s.Require().NotNil(addedPermission)
+	s.Require().Equal(communities.TokenPermissionAdditionPending, addedPermission.State)
+
+	messengerReponse, err := s.john.PromoteSelfToControlNode(community.ID())
+
+	s.Require().NoError(err)
+	s.Require().Len(messengerReponse.Communities(), 1)
+
+	tokenPermissions := messengerReponse.Communities()[0].TokenPermissions()
+	s.Require().Len(tokenPermissions, 2)
+
+	tokenPermissionsMap := make(map[protobuf.CommunityTokenPermission_Type]struct{}, len(tokenPermissions))
+	for _, t := range tokenPermissions {
+		tokenPermissionsMap[t.Type] = struct{}{}
+	}
+
+	s.Require().Len(tokenPermissionsMap, 2)
+	s.Require().Contains(tokenPermissionsMap, protobuf.CommunityTokenPermission_BECOME_TOKEN_OWNER)
+	s.Require().Contains(tokenPermissionsMap, protobuf.CommunityTokenPermission_BECOME_ADMIN)
+
+	for _, v := range tokenPermissions {
+		s.Require().Equal(communities.TokenPermissionApproved, v.State)
+	}
+}
+
+func (s *MessengerCommunitiesSignersSuite) TestWithoutMintedOwnerTokenMakingDeviceControlNodeIsBlocked() {
+	community := s.createCommunity(s.john)
+
+	// Make sure there is no control node
+	s.Require().False(common.IsPubKeyEqual(community.ControlNode(), &s.john.identity.PublicKey))
+
+	response, err := s.john.PromoteSelfToControlNode(community.ID())
+	s.Require().Nil(response)
+	s.Require().NotNil(err)
+	s.Require().Error(err, "Owner token is needed")
+}

--- a/protocol/messenger_communities.go
+++ b/protocol/messenger_communities.go
@@ -71,6 +71,18 @@ const (
 	maxChunkSizeBytes    = 1500000
 )
 
+const (
+	ErrOwnerTokenNeeded                     = "Owner token is needed" // #nosec G101
+	ErrMissingCommunityID                   = "CommunityID has to be provided"
+	ErrForbiddenProfileOrWatchOnlyAccount   = "Cannot join a community using profile chat or watch-only account"
+	ErrSigningJoinRequestForKeycardAccounts = "Signing a joining community request for accounts migrated to keycard must be done with a keycard"
+	ErrNotPartOfCommunity                   = "Not part of the community"
+	ErrNotAdminOrOwner                      = "Not admin or owner"
+	ErrSignerIsNil                          = "Signer can't be nil"
+	ErrSyncMessagesSentByNonControlNode     = "Accepted/requested to join sync messages can be send only by the control node"
+	ErrReceiverIsNil                        = "Receiver can't be nil"
+)
+
 type FetchCommunityRequest struct {
 	// CommunityKey should be either a public or a private community key
 	CommunityKey    string       `json:"communityKey"`
@@ -1175,7 +1187,7 @@ func (m *Messenger) generateCommunityRequestsForSigning(memberPubKey string, com
 
 func (m *Messenger) GenerateJoiningCommunityRequestsForSigning(memberPubKey string, communityID types.HexBytes, addressesToReveal []string) ([]account.SignParams, error) {
 	if len(communityID) == 0 {
-		return nil, errors.New("communityID has to be provided")
+		return nil, errors.New(ErrMissingCommunityID)
 	}
 	return m.generateCommunityRequestsForSigning(memberPubKey, communityID, addressesToReveal, false)
 }
@@ -1200,7 +1212,7 @@ func (m *Messenger) SignData(signParams []account.SignParams) ([]string, error) 
 		}
 
 		if account.Chat || account.Type == accounts.AccountTypeWatch {
-			return nil, errors.New("cannot join a community using profile chat or watch-only account")
+			return nil, errors.New(ErrForbiddenProfileOrWatchOnlyAccount)
 		}
 
 		keypair, err := m.settings.GetKeypairByKeyUID(account.KeyUID)
@@ -1209,7 +1221,7 @@ func (m *Messenger) SignData(signParams []account.SignParams) ([]string, error) 
 		}
 
 		if keypair.MigratedToKeycard() {
-			return nil, errors.New("signing a joining community request for accounts migrated to keycard must be done with a keycard")
+			return nil, errors.New(ErrSigningJoinRequestForKeycardAccounts)
 		}
 
 		verifiedAccount, err := m.accountsManager.GetVerifiedWalletAccount(m.settings, param.Address, param.Password)
@@ -1419,7 +1431,7 @@ func (m *Messenger) EditSharedAddressesForCommunity(request *requests.EditShared
 	}
 
 	if !community.HasMember(m.IdentityPublicKey()) {
-		return nil, errors.New("not part of the community")
+		return nil, errors.New(ErrNotPartOfCommunity)
 	}
 
 	revealedAddresses := make([]gethcommon.Address, 0)
@@ -2332,7 +2344,7 @@ func (m *Messenger) SetCommunityShard(request *requests.SetCommunityShard) (*Mes
 	}
 
 	if !community.IsControlNode() {
-		return nil, errors.New("not admin or owner")
+		return nil, errors.New(ErrNotAdminOrOwner)
 	}
 
 	// Reset the community private key
@@ -2411,7 +2423,7 @@ func (m *Messenger) SetCommunityStorenodes(request *requests.SetCommunityStoreno
 		return nil, err
 	}
 	if !community.IsControlNode() {
-		return nil, errors.New("not admin or owner")
+		return nil, errors.New(ErrNotAdminOrOwner)
 	}
 
 	if err := m.communityStorenodes.UpdateStorenodesInDB(request.CommunityID, request.Storenodes, 0); err != nil {
@@ -3264,7 +3276,7 @@ func (m *Messenger) HandleCommunityShardKey(state *ReceivedMessageState, message
 
 	signer := state.CurrentMessageState.PublicKey
 	if signer == nil {
-		return errors.New("signer can't be nil")
+		return errors.New(ErrReceiverIsNil)
 	}
 
 	err = m.handleCommunityShardAndFiltersFromProto(community, message)
@@ -3324,7 +3336,7 @@ func (m *Messenger) handleCommunityShardAndFiltersFromProto(community *communiti
 
 func (m *Messenger) handleCommunityPrivilegedUserSyncMessage(state *ReceivedMessageState, signer *ecdsa.PublicKey, message *protobuf.CommunityPrivilegedUserSyncMessage) error {
 	if signer == nil {
-		return errors.New("signer can't be nil")
+		return errors.New(ErrSignerIsNil)
 	}
 
 	community, err := m.communitiesManager.GetByID(message.CommunityId)
@@ -3337,7 +3349,7 @@ func (m *Messenger) handleCommunityPrivilegedUserSyncMessage(state *ReceivedMess
 	// CONTROL_NODE were sent by a control node
 	isControlNodeMsg := common.IsPubKeyEqual(community.ControlNode(), signer)
 	if !isControlNodeMsg {
-		return errors.New("accepted/requested to join sync messages can be send only by the control node")
+		return errors.New(ErrSyncMessagesSentByNonControlNode)
 	}
 
 	err = m.communitiesManager.ValidateCommunityPrivilegedUserSyncMessage(message)
@@ -3373,7 +3385,7 @@ func (m *Messenger) HandleCommunityPrivilegedUserSyncMessage(state *ReceivedMess
 
 func (m *Messenger) sendSharedAddressToControlNode(receiver *ecdsa.PublicKey, community *communities.Community) (*communities.RequestToJoin, error) {
 	if receiver == nil {
-		return nil, errors.New("receiver can't be nil")
+		return nil, errors.New(ErrReceiverIsNil)
 	}
 
 	if community == nil {
@@ -4499,6 +4511,10 @@ func (m *Messenger) PromoteSelfToControlNode(communityID types.HexBytes) (*Messe
 
 	if err != nil {
 		return nil, err
+	}
+
+	if !communities.HasTokenOwnership(community.Description()) {
+		return nil, errors.New(ErrOwnerTokenNeeded)
 	}
 
 	changes, err := m.communitiesManager.PromoteSelfToControlNode(community, clock)

